src/OrchardCore.Modules/OrchardCore.ContentLocalization/ContentRequestCultureProvider.cs

```diff
@@ -0,0 +1,36 @@
+using System;
+using Microsoft.Extensions.DependencyInjection;
+using System.Threading.Tasks;
+using YesSql;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Localization;
+using OrchardCore.Autoroute.Services;
+using OrchardCore.ContentLocalization.Records;
+using OrchardCore.ContentManagement;
+
+namespace OrchardCore.ContentLocalization
+{
+    /// <summary>
+    /// RequestCultureProvider that automatically sets the Culture of a request from the LocalizationPart.Culture property.
+    /// </summary>
+    public class ContentRequestCultureProvider : RequestCultureProvider
+    {
+        public override async Task<ProviderCultureResult> DetermineProviderCultureResult(HttpContext httpContext)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+            var culturePickerService = httpContext.RequestServices.GetService<IContentCulturePickerService>();
+            
+            var path = httpContext.Request.PathBase + httpContext.Request.Path + httpContext.Request.QueryString;
+
+            var culture = await culturePickerService.GetCultureFromRoute(path);
```

Comment:
This probably has a big perf impact. Some type of In-Memory cache should probably be used here.